addons/loyalty/models/loyalty_history.py

```diff
@@ -1,28 +1,191 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import fields, models
+from datetime import timedelta
+
+from odoo import api, fields, models
 
 
 class LoyaltyHistory(models.Model):
     _name = 'loyalty.history'
     _description = "History for Loyalty cards and Ewallets"
     _order = 'id desc'
 
-    card_id = fields.Many2one(comodel_name='loyalty.card', required=True, index=True, ondelete='cascade')
+    card_id = fields.Many2one(
+        comodel_name='loyalty.card',
+        ondelete='cascade',
+        readonly=True,
+        required=True,
+        index=True,
+    )
+    program_type = fields.Selection(related='card_id.program_type')
     company_id = fields.Many2one(related='card_id.company_id')
 
     description = fields.Text(required=True)
 
-    issued = fields.Float()
-    used = fields.Float()
+    issued = fields.Float(readonly=True)
+    available_issued_points = fields.Float(string='Available', readonly=True)
+    used = fields.Float(
+        help="Indicates the number of points claimed on a sale order;"
+             " not necessarily deducted from the points issued on the same line.",
+        readonly=True,
+    )
+    active = fields.Boolean(default=True)
+    expiration_date = fields.Date(string='Expiration')
 
     order_model = fields.Char(readonly=True)
     order_id = fields.Many2oneReference(model_field='order_model', readonly=True)
 
+    def _get_sorted_history_lines(self, lines):
+        """
+        Sorts loyalty history lines by redemption priority.
+
+        Points are ordered by:
+            1. Nearest expiration date (soonest first).
+            2. Earlier creation order, if expiration dates are equal.
+            3. Points without an expiration date are placed last.
+        """
+        return lines.sorted(
+            key=lambda line: (
+                line.expiration_date is False,
+                line.expiration_date,
+                line.id,
+            ),
+        )
+
     def _get_order_portal_url(self):
         self.ensure_one()
         return False
 
     def _get_order_description(self):
         self.ensure_one()
         return self.env[self.order_model].browse(self.order_id).display_name
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        today = fields.Date.today()
+        for vals in vals_list:
+            # The history line is archived immediately as it has no redeemable points.
+            if vals.get('available_issued_points') == 0:
+                vals['active'] = False
+                continue
+            card = self.env['loyalty.card'].browse(vals.get('card_id'))
+            if expire_after := card.program_id.expire_after:
+                vals['expiration_date'] = today + timedelta(days=expire_after)
+        return super().create(vals_list)
+
+    def write(self, vals):
+        result = super().write(vals)
+        today = fields.Date.today()
+        if 'expiration_date' in vals:
```

Comment:
I don't think it should be allowed to set the expiry date in the past. There is a bugfix for expiration for loyalty card, so it makes sense to have the same contrain here.